src/api/schema/filter.rs

```diff
@@ -0,0 +1,188 @@
+use super::components::{source, ComponentType};
+use async_graphql::{InputObject, InputType};
+
+/// Takes an &Option<bool> and returns early if false
+#[macro_export]
+macro_rules! filter_check {
+    ($($match:expr),+) => {
+        $(
+            if matches!($match, Some(t) if !t) {
+                return false;
+            }
+        )+
+    }
+}
+
+#[derive(Default, InputObject)]
+/// Filter for String values
+pub struct StringFilter {
+    pub equals: Option<String>,
+    pub not_equals: Option<String>,
+    pub contains: Option<String>,
+    pub not_contains: Option<String>,
+    pub starts_with: Option<String>,
+    pub ends_with: Option<String>,
+}
+
+impl StringFilter {
+    pub fn filter_value(&self, value: &str) -> bool {
+        filter_check!(
+            // Equals
+            self.equals.as_ref().map(|s| value.eq(s)),
+            // Not equals
+            self.not_equals.as_ref().map(|s| !value.eq(s)),
+            // Contains
+            self.contains.as_ref().map(|s| value.contains(s)),
+            // Does not contain
+            self.not_contains.as_ref().map(|s| !value.contains(s)),
+            // Starts with
+            self.starts_with.as_ref().map(|s| value.starts_with(s)),
+            // Ends with
+            self.ends_with.as_ref().map(|s| value.ends_with(s))
+        );
+        true
+    }
+}
+
+#[derive(InputObject)]
+#[graphql(concrete(name = "ComponentTypeFilter", params(ComponentType)))]
+#[graphql(concrete(name = "SourceOutputTypeFilter", params(source::SourceOutputType)))]
```

Comment:
Per https://github.com/async-graphql/async-graphql/issues/387, generic `InputObject`s are supported, but the concrete implementations are required to be defined alongside it.

This is a little unfortunate for cross-cutting concerns, as it prevents ad-hoc implementations as would generally be the case with generics. On the plus side, it makes it obvious what implements it in one place.